dex/price-discovery/src/user_actions/redeem.rs

```diff
@@ -0,0 +1,124 @@
+use crate::user_actions::user_deposit_withdraw::INVALID_PAYMENT_ERR_MSG;
+
+multiversx_sc::imports!();
+
+#[multiversx_sc::module]
+pub trait RedeemModule:
+    super::user_deposit_withdraw::UserDepositWithdrawModule
+    + crate::common_storage::CommonStorageModule
+    + crate::events::EventsModule
+    + crate::phase::PhaseModule
+    + crate::redeem_token::RedeemTokenModule
+    + multiversx_sc_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+{
+    /// After all phases have ended,
+    /// users can withdraw their fair share of launched tokens,
+    /// while the owner can withdraw the accepted tokens.
+    #[payable("*")]
+    #[endpoint]
+    fn redeem(&self) -> EgldOrEsdtTokenPayment {
+        self.require_redeem_token_setup_complete();
+
+        let phase = self.get_current_phase();
+        self.require_redeem_allowed(&phase);
+
+        let caller = self.blockchain().get_caller();
+        let owner = self.blockchain().get_owner_address();
+        if caller == owner {
+            self.require_no_payment();
+
+            let redeemed_tokens = self.owner_redeem(&caller);
+
+            self.emit_redeem_event(
+                None,
+                &BigUint::zero(),
+                &redeemed_tokens.token_identifier,
+                &redeemed_tokens.amount,
+            );
+
+            return redeemed_tokens;
+        }
+
+        let (payment_token, payment_amount) = self.call_value().single_fungible_esdt();
+        let bought_tokens = self.user_redeem(&caller, &payment_token, &payment_amount);
+
+        self.emit_redeem_event(
+            Some(&payment_token),
+            &payment_amount,
+            &bought_tokens.token_identifier,
+            &bought_tokens.amount,
+        );
+
+        bought_tokens
+    }
+
+    fn owner_redeem(&self, owner: &ManagedAddress) -> EgldOrEsdtTokenPayment {
+        let launched_token_supply = self.launched_token_balance().get();
+        require!(
+            launched_token_supply > 0,
+            "May not withdraw tokens as launched tokens were not deposited"
+        );
+
+        let accepted_token_id = self.accepted_token_id().get();
+        let accepted_token_balance = self.accepted_token_balance().get();
+        self.send()
+            .direct(owner, &accepted_token_id, 0, &accepted_token_balance);
+
+        EgldOrEsdtTokenPayment::new(accepted_token_id, 0, accepted_token_balance)
+    }
+
+    fn user_redeem(
+        &self,
+        user: &ManagedAddress,
+        payment_token: &TokenIdentifier,
+        payment_amount: &BigUint,
+    ) -> EgldOrEsdtTokenPayment {
+        require!(self.is_user_whitelisted(user), "User not whitelisted");
```

Comment:
Why this check? The user should not have redeem tokens if he is not whitelisted. We should either move this check to the userDeposit endpoint or remove it altogether. Otherwise, it might backfire, users buying redeem tokens and not being able to redeem the launch token or even his initial tokens.

Reply:
Alright.